### ISSUE TITLE
wrong lifetime error with nested flat_map + map

### ISSUE THREAD

**Issue by durka** [rust-lang/rust#48805]
This code

```rust
(1..2).flat_map(
    |x| {
        (1..2).map(|y| x * y)
    });
```

produces the following error:

```
error[E0597]: `x` does not live long enough
 --> src/main.rs:4:28
  |
4 |             (1..2).map(|y| x * y)
  |                        --- ^ borrowed value does not live long enough
  |                        |
  |                        capture occurs here
5 |         });
  |         -- borrowed value needs to live until here
  |         |
  |         borrowed value only lives until here
```

This is clearly nonsense as `x` is a `u64`, not a borrowed value.

`flat_map` often produces bonkers lifetime errors but this one has me stumped. What is the underlying error (if any) and why is it getting reported this way?
